Display transaction notes on mobile, fixes #1764

DarkWolfSLV opened the pull request:
All credits go to [a-gradina](https://github.com/a-gradina), I just completed his pull request. 
This change will display the transaction notes on the mobile version of Actual. 

![image](https://github.com/user-attachments/assets/d69de5a4-52db-4e58-820a-29d0568dce9c)

Fixes issue [1764](https://github.com/actualbudget/actual/issues/1764)

youngcw commented:
I don't love the inconsistent alignment between with notes and without notes transactions.  But it would be good to get other opinions

youngcw commented:
if you want a full review you need to remove the wip label

matt-fidd commented:
> I don't love the inconsistent alignment between with notes and without notes transactions. But it would be good to get other opinions

I agree, and it already feels like real estate is limited on mobile sometimes. In fairness though, I also don't have any better ideas.

DarkWolfSLV commented:
> > I don't love the inconsistent alignment between with notes and without notes transactions. But it would be good to get other opinions
> 
> I agree, and it already feels like real estate is limited on mobile sometimes. In fairness though, I also don't have any better ideas.

Can you elaborate on the alignment issue? I'm assuming the top/bottom space for the no notes (2 lines) vs with notes (3 lines). 
Opt 1 and Opt 2 below are the other alternatives, but I don't like them because we are showing 3 things, we should list 3 things and that creates another complication if any of the fields (Payee or Category) are long.

Opt 3 is the same as the current solution alignment-wise, so I don't think it help us much. 

![image](https://github.com/user-attachments/assets/2e500f9e-a113-46f5-b287-3513836e8bba)

joel-jeremy commented:
The UI feels a bit cramped IMO.

I wonder if we can add a expand/collapse button to show/hide more info about a transaction e.g. Notes, Running Balance, Splits Overview, etc

Teprifer commented:
I think it's fine, I prefer opt 3 above as mentioned previously as it allows for longer notes and the category isn't changing position within each add you scroll.

I think the changing height is quieter with opt 3 as well because of that. 


Using the test budget, there seems to be a crash when adding a note to a split transaction. If you're not able to reproduce let me know and I'll grab the console error when I'm at my PC.

DarkWolfSLV commented:
> I think it's fine, I prefer opt 3 above as mentioned previously as it allows for longer notes and the category isn't changing position within each transaction as you scroll.
> 
> I think the changing height is quieter with opt 3 as well because of that.
> 
> Using the test budget, there seems to be a crash when adding a note to a split transaction. If you're not able to reproduce let me know and I'll grab the console error when I'm at my PC.

I will try to move it below the category, but I doubt I can solve the slip bug. I might need more help with that.

Teprifer commented:
> > I think it's fine, I prefer opt 3 above as mentioned previously as it allows for longer notes and the category isn't changing position within each transaction as you scroll.
> > I think the changing height is quieter with opt 3 as well because of that.
> > Using the test budget, there seems to be a crash when adding a note to a split transaction. If you're not able to reproduce let me know and I'll grab the console error when I'm at my PC.
> 
> I will try to move it below the category, but I doubt I can solve the slip bug. I might need more help with that.

Ah it may not be related to your work at all, saw mention of this in discord: https://github.com/actualbudget/actual/pull/4186

youngcw commented:
> The UI feels a bit cramped IMO.
> 
> I wonder if we can add a functionality to show/hide more info about a transaction e.g. Notes, Running Balance, Splits Overview, etc. This could probably be done by:
> 
> 1. Adding a expand/collapse (and a expand all/collapse all button maybe?) button to show/hide this info
> 2. Changing the behavior of clicking a transaction to open a modal with info about the transaction and add a button on the modal to `Edit transaction`
>    
>    * The modal could then be swiped left/right to go to previous/next transaction to quickly browse transactions
> 3. In transaction edit screen, add a previous/next button (or swipe) to go to previous/next transaction (based on the page you opened the transaction from e.g. account transactions page/category transactions page)
> 
> Just throwing ideas out there

I like option 1 here as a long term solution

matt-fidd commented:
I'm a fan of the expand/collapse too

DarkWolfSLV commented:
Hi everyone! 
I updated the code to show the notes at the end. I tried to slice the notes but it crashed every time. So I left it as it was before with just `{notes}` -- I was using this code if someone wants to try it `{notes.length > 30 ? '${notes.slice(0, 30)}...' : notes}`

I know a lot of you would like the hide feature, but I do not know how to implement it - can we merge this as v1 once we fix the slice? 

@Teprifer FYI - With this addition, split transactions do not show notes.

![image](https://github.com/user-attachments/assets/0e8137f6-1f8c-4ec2-8a9d-be366b7049fd)

youngcw commented:
This is what Im getting currently.  Its not on its own line
![image](https://github.com/user-attachments/assets/25857e2f-685c-4f09-894c-270c366251bd)

youngcw commented:
That looks decent.  Until we add an expandable option this should work.

joel-jeremy commented:
I'm getting this:
![Screenshot_20250120-130334](https://github.com/user-attachments/assets/a98a8367-ef04-45e1-abf6-9feffbfdad7d)

The text needs to be truncated and replaced with an ellipsis when it gets too long

iv-rvalenzuela commented:
@joel-jeremy I tried with `{notes.length > 30 ? '${notes.slice(0, 30)}...' : notes}` and that was giving me an error - I was hoping someone had an idea of why to try to correct it. FYI, in the code, I was using the backticks (`) instead of single quotes (').

joel-jeremy commented:
> @joel-jeremy I tried with `{notes.length > 30 ? '${notes.slice(0, 30)}...' : notes}` and that was giving me an error - I was hoping someone had an idea of why to try to correct it. FYI, in the code, I was using the backticks (`) instead of single quotes (').

Hmm, weird. The TextOneLine is supposed to auto truncate since it has `textOverflow: 'ellipsis'` and `overflow: 'hidden'` styles

joel-jeremy commented on packages/desktop-client/src/components/mobile/transactions/TransactionListItem.tsx:
```diff
@@ -195,57 +196,69 @@ export function TransactionListItem({
+                        {prettyCategory || 'Uncategorized'}
+                      </TextOneLine>
+                    </View>
                     <TextOneLine
```
This should be after [line 264](https://github.com/actualbudget/actual/pull/4159/files#diff-f558919867b6ae90e9d47cb380bf61308cc8f636a89c98d52d1fcb7738f6d605R264). After moving this, update [line 166](https://github.com/actualbudget/actual/pull/4159/files#diff-f558919867b6ae90e9d47cb380bf61308cc8f636a89c98d52d1fcb7738f6d605R166) with `style={{ flex: 1}}`

DarkWolfSLV replied:
Hi @joel-jeremy this comment wasn't super clear to me, but I think I got it.
Please let me know what you think - The text is also getting truncated automatically now. 

Thank you for all the help. 
![image](https://github.com/user-attachments/assets/212939af-6d0e-40e1-9edc-062bfa7f8469)

joel-jeremy commented on packages/desktop-client/src/components/mobile/transactions/TransactionListItem.tsx:
```diff
@@ -195,57 +196,69 @@ export function TransactionListItem({
                 ) : (
                   <View
                     style={{
-                      flexDirection: 'row',
-                      alignItems: 'center',
+                      flexDirection: 'column',
+                      alignItems: 'flex-start',
                       marginTop: 3,
                     }}
                   >
```
This new View is not needed. Please see other comment below.

joel-jeremy commented on packages/desktop-client/src/components/mobile/transactions/TransactionListItem.tsx:
```diff
@@ -249,6 +250,18 @@ export function TransactionListItem({
                     </TextOneLine>
                   </View>
                 )}
+                <TextOneLine
+                  style={{
+                    fontSize: 11,
+                    marginTop: 1,
+                    fontWeight: '400',
+                    color: theme.tableText,
+                    textAlign: 'left',
+                    opacity: 0.75,
+                  }}
+                >
+                  {notes}
+                </TextOneLine>
```
This is a good suggestion

joel-jeremy reviewed (approved):
LGTM!
